File: crates/upmd-runner/src/languages/typescript/state.ts
// State capture functionality for TypeScript
// This code is injected into user scripts for experimental state capture

import * as fs from 'fs';
import * as process from 'process';

function upmdCaptureState(): void {
    upmdWriteState();
}

function upmdStateEscape(value: string): string {
    return value
        .replace(/\\/g, '\\\\')
        .replace(/"/g, '\\"')
        .replace(/\n/g, '\\n')
        .replace(/\r/g, '\\r')
        .replace(/\t/g, '\\t');
}

function upmdWriteState(): void {
    const stateFifo = process.env.UPMD_STATE_FIFO;
    if (!stateFifo) return;

    try {
        const lines = [
            'version 1',
            `cwd "${upmdStateEscape(process.cwd())}"`,
        ];
        for (const [key, value] of Object.entries(process.env)) {
            lines.push(`env "${upmdStateEscape(key)}" "${upmdStateEscape(value || '')}"`);
        }
        fs.writeFileSync(stateFifo, `${lines.join('\n')}\n`);
    } catch (error) {
        // Silently ignore errors
    }
}

File: crates/upmd-runner/src/languages/python/state.py
# State capture functionality for Python
# This code is injected into user scripts for experimental state capture

import os

def upmd_capture_state():
    """Capture environment variables and current working directory."""
    upmd_write_state()

def upmd_state_escape(value):
    return (
        value
        .replace('\\', '\\\\')
        .replace('"', '\\"')
        .replace('\n', '\\n')
        .replace('\r', '\\r')
        .replace('\t', '\\t')
    )

def upmd_write_state():
    """Write environment variables and cwd to FIFO as upmd state v1."""
    state_fifo = os.environ.get('UPMD_STATE_FIFO', '/dev/null')
    try:
        with open(state_fifo, 'w') as f:
            f.write('version 1\n')
            f.write(f'cwd "{upmd_state_escape(os.getcwd())}"\n')
            for key, value in os.environ.items():
                f.write(f'env "{upmd_state_escape(key)}" "{upmd_state_escape(value)}"\n')
    except Exception:
        pass  # Silently ignore state capture errors
